pkg/controllers/vdb/readonlyonlineupgrade_reconciler.go

```diff
@@ -454,17 +457,59 @@ func (o *ReadOnlyOnlineUpgradeReconciler) drainSubcluster(ctx context.Context, s
 		return ctrl.Result{}, err
 	}
 
-	if img != o.Vdb.Spec.Image {
-		scName := sts.Labels[vmeta.SubclusterNameLabel]
-		o.Log.Info("rerouting client traffic from subcluster", "name", scName)
-		if err := o.routeClientTraffic(ctx, scName, true); err != nil {
-			return ctrl.Result{}, err
-		}
+	if img == o.Vdb.Spec.Image {
+		return ctrl.Result{}, nil
+	}
+	scName := sts.Labels[vmeta.SubclusterNameLabel]
+	o.Log.Info("rerouting client traffic from subcluster", "name", scName)
+	if err := o.routeClientTraffic(ctx, scName, true); err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	o.Log.Info("starting check for active connections in subcluster", "name", scName)
+	return o.runSubclusterDrain(ctx, scName)
+}
+
+func (o *ReadOnlyOnlineUpgradeReconciler) runSubclusterDrain(ctx context.Context, scName string) (ctrl.Result, error) {
+	drainSubclusterAnnotation := "vertica.com/drain-start-time-" + scName
+	pfs := o.PFacts.FindPodsBySubcluster(scName)
 
-		o.Log.Info("starting check for active connections in subcluster", "name", scName)
-		return o.Manager.isSubclusterIdle(ctx, o.PFacts, scName)
+	// If no active connections, return
+	if res, err := o.Manager.isSubclusterIdle(ctx, o.PFacts, scName); !verrors.IsReconcileAborted(res, err) {
+		return ctrl.Result{}, removeDrainStartAnnotation(ctx, o.Vdb, o.VRec, drainSubclusterAnnotation)
 	}
-	return ctrl.Result{}, nil
+
+	timeoutInt := o.Vdb.GetActiveConnectionsDrainSeconds()
+	// If timeout is zero, proceed
+	if timeoutInt == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	drainStartTimeStr, found := o.Vdb.Annotations[drainSubclusterAnnotation]
+	// If drain start time annotation is not set, we set it and requeue after 1s
+	if !found {
+		o.Log.Info("Starting draining before upgrade")
+		return ctrl.Result{RequeueAfter: 1 * time.Second}, setDrainStartAnnotation(ctx, o.Vdb, o.VRec, drainSubclusterAnnotation)
+	}
+	var drainStartTime time.Time
+	drainStartTime, err := time.Parse(time.RFC3339, drainStartTimeStr)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	elapsed := time.Since(drainStartTime)
+	timeout := time.Duration(timeoutInt) * time.Second
+	// If timeout has expired, we move on to the next reconciler (DBRemoveNodeReconciler|DBRemoveSubclusterReconciler)
+	o.Log.Info("Draining in progress", "start", drainStartTime, "elapsed", elapsed, "timeout", timeout)
+	if elapsed >= timeout {
+		o.VRec.Eventf(o.Vdb, corev1.EventTypeWarning, events.DrainSubclusterTimeout,
+			"Timed out while draining connections for subcluster '%s'; killing connections", scName)
+		o.Log.Info("Draining timeout has expired")
+		killConnectionsToPendingDeletePods(ctx, o.VRec, o.PFacts.PRunner, pfs)
```

Comment:
This will make us try to kill connections in every pod in the subcluster (even pods with no active connections) and that's expensive. I think we should have another function like isSubclusterIdle (with a slightly different query) that will return the list of session ids in the subcluster. If the list is empty, we know there are no active connections in the subcluster. If non-empty, we can later reuse that sesssion ids list to kill the sessions.